File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let low = 0;
    let maxLen = 0;
    let map = new Map();
    for (let high=0; high < s.length; high++) {
        let right = s[high];
        map.set(right, (map.get(right) || 0) + 1);
        while (map.size < (high - low + 1)) {
            let left = s[low];
            map.set(left, map.get(left) - 1);
            if (map.get(left) === 0) map.delete(left);
            low++;
        }
        maxLen = Math.max(maxLen, high - low + 1);
    }
    return maxLen;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int low = 0;
        int maxLen = 0;
        map<char, int> mp;

        for (int high=0; high < s.size(); high++) {
            char right = s[high];
            mp[right]++;

            while (mp.size() < (high - low + 1)) {
                char left = s[low];
                mp[left]--;
                if (mp[left] == 0) mp.erase(left);
                low++;
            }
            maxLen = max(maxLen, high - low + 1);
        }
        return maxLen;
    }
};